### PR TITLE
Relax upper bounds

### PR DESCRIPTION
**Pull request by phadej**
Is it ok to add released packages to Stackage, so we get pinged when bounds are restrictive? `hackage-security` is anyhow there, as it's depended by `cabal-install`.

### CODE REVIEW

**Comment by ezyang** [comment]
Hmm, is there a way to do this patch with less CPP?

**Comment by phadej** [comment]
If it's ok to have `http-client >= 0.5`

**Comment by ezyang** [comment]
I don't really have the final say for a package like this, but my preference is that we create a Compat package that centralizes compatibility wrappers around API differences between the old and new versions.

**Comment by phadej** [comment]
It'S not realised example package, so i'd rather not over-engineer. I'll be ok with CPP or tighter lower bound, otherwise the value of an example degrares

**Comment by ezyang** [comment]
Oops, I didn't realize it was an example package. In that case, updating the version bound to the newest one seems like a clear winner.

**Comment by edsko** [comment]
Moving to http-client-0.5 for the -http-client bindings seems perfectly fine to me, so let's get rid of the CPP and just change the bounds?

**Comment by edsko** [comment]
I'm okay twith leaving the bounds too by the way, actually. That package is meant to show how http-client can be used together with hackage-security in case, say, Stackage wanted to use it. But I don't know if it's _actually_ being used. With the CPP it would demonstrate I suppose how either version works, and LTS 7 at least still uses http-client 0.4. @dcoutts , opinions?

**Comment by hvr** [comment]
My opinion here is mostly to encourage reducing CPP usage :-)

**Comment by phadej** [comment]
[LTS-7](https://www.stackage.org/lts-7.0) series is released on 2016-09-14. To Stackage's defense, they didn't tried to bump too much until then, but cut GHC-8.0.1 based series from what they got. Right after that, the blocking packages were removed, so e.g. http-client-0.5 is in [nightly-2016-09-15](https://www.stackage.org/nightly-2016-09-15)

This diff is an interesting datapoint: https://www.stackage.org/diff/nightly-2016-09-14/nightly-2016-09-15 (there was packages blocking on `aeson-1`, `http-client-0.5` and `haskell-src-exts-0.18`)

---

But, remove CPP or leave it?

**Comment by edsko** [comment]
_If_ Stackage decides to support `hackage-security`, would they want backwards compatibility with http-client-0.4? If the answer is no, then let's get rid of the CPP. If the answer is yes, then let's keep it.

**Comment by phadej** [comment]
newest release of `stack` depends on `http-client >= 0.5`. But to clarify, the `http-client` dependency is in unreleased example code `hackage-security-http-client`, not `hackage-security`.

I'll amend this PR shortly

**Comment by phadej** [comment]
FYI

https://github.com/commercialhaskell/stack/pull/2940

**Comment by phadej** [comment]
Please merge, or say I can remove CPP in favour of `http-client >= 0.5`.

So small change "stuck in committee"

**Comment by edsko** [comment]
The only purpose of hackage-security-http-client from the very beginning was to make it easier for stack to use hackage-security. If backwards compatibility doesn't matter to stack, let's get rid of the CPP.

Incidentally, overlap with #180?

**Comment by phadej** [comment]
I'll integrate @snoyberg's patch into this, the `makeHttpLib`.

**Comment by phadej** [comment]
Integrated #180 changes.

**Comment by phadej** [comment]
`http-client-0.5.` has `bytestring >= 0.10`, so `--constraint="bytestring installed"`, doesn't work for GHC-7.4, as it has `bytestring-0.9.2.1` https://ghc.haskell.org/trac/ghc/wiki/Commentary/Libraries/VersionHistory

So it's either "keep CI simple" or "support http-client-0.4". Ping @edsko.

**Comment by edsko** [comment]
The `bytestring installed` constraint as well as support for older ghc is important for the main `hackage-security` library, but neither the contraint nor support for older ghc is important for the `-http-client` library.

**Comment by phadej** [comment]
Finally green. I miss `shelly` or something like that.

**Comment by phadej** [comment]
ping @edsko

File: .travis.yml [hunk 1]
```diff
@@ -35,6 +35,7 @@ before_install:
  - HC=${CC}
  - unset CC
  - PATH=/opt/ghc/bin:/opt/ghc-ppa-tools/bin:$PATH
+ - if [ $GHCVER = "7.4.2" ]; then mv cabal.ghc-7.4.2.project cabal.project; fi
```
Comment (hvr) [inline .travis.yml]: This most likely isn't needed and we should avoid it. While Stack may have problems coping with that configuration I honestly doubt that Cabal has any issues with this. Have you run into an actual problem? Also it'd defeat the purpose of CI if we exclude configurations even though they're declared to be compatible via `build-depends`.

**Comment by phadej** [comment]
Green again.

File: .travis.yml [hunk 2]
```diff
@@ -51,10 +51,18 @@ script:
  - for PKG in "${PKGS[@]}"; do
      cd "$PKG"; cabal sdist --output-directory="../sdists/$PKG" || break; cd ..;
    done
- - cp -v cabal.project sdists/
  - cd sdists/
-# build all packages and run testsuite
+
+ # first build just hackage-security with installed constraints, with and without tests.
+ # silly yaml, seeing : colon
```
Comment (hvr) [inline .travis.yml]: yeah... just one of many YAML silliness ;-)

**Comment by hvr** [comment]
I'll try to code-review to make sure that the relaxation to the new major versions of `optparse-applicative` and `directory` are sound; I think I can safely assume that `http-client-0.5` compatibility is sound since it was a patch contributed by `http-client'`s author

**Pull request merged**